src/python/pants/backend/python/lint/isort/rules_integration_test.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import List, Optional, Tuple
+
+from pants.backend.python.lint.isort.rules import IsortSetup, fmt, lint, setup_isort
+from pants.backend.python.lint.isort.subsystem import Isort
+from pants.backend.python.rules.download_pex_bin import download_pex_bin
+from pants.backend.python.rules.pex import CreatePex, create_pex
+from pants.backend.python.subsystems.python_native_code import (
+  PythonNativeCode,
+  create_pex_native_build_environment,
+)
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.backend.python.subsystems.subprocess_environment import (
+  SubprocessEnvironment,
+  create_subprocess_encoding_environment,
+)
+from pants.backend.python.targets.formattable_python_target import FormattablePythonTarget
+from pants.build_graph.address import Address
+from pants.engine.fs import Digest, FileContent, InputFilesContent, Snapshot
+from pants.engine.legacy.structs import TargetAdaptor
+from pants.engine.rules import RootRule
+from pants.engine.selectors import Params
+from pants.rules.core.fmt import FmtResult
+from pants.rules.core.lint import LintResult
+from pants.source.wrapped_globs import EagerFilesetWithSpec
+from pants.testutil.subsystem.util import global_subsystem_instance, init_subsystems
+from pants.testutil.test_base import TestBase
+from pants.util.collections import assert_single_element
+
+
+class IsortIntegrationTest(TestBase):
+
+  good_source = FileContent(path="test/good.py", content=b'from colors import blue, green\n')
+  bad_source = FileContent(path="test/bad.py", content=b'from colors import green, blue\n')
+  fixed_bad_source = FileContent(path="test/bad.py", content=b'from colors import blue, green\n')
+
+  @classmethod
+  def rules(cls):
+    return (
+      *super().rules(),
+      fmt,
+      lint,
+      setup_isort,
+      create_pex,
+      create_subprocess_encoding_environment,
+      create_pex_native_build_environment,
+      download_pex_bin,
+      RootRule(CreatePex),
+      RootRule(FormattablePythonTarget),
+      RootRule(Isort),
+      RootRule(IsortSetup),
+      RootRule(PythonSetup),
+      RootRule(PythonNativeCode),
+      RootRule(SubprocessEnvironment),
+    )
```

Comment:
This—and the corresponding imports—pains me. I would like us to be able to simply say something like `isort.register.rules()`. The reason we cannot do this is that we need any `optionable_rule` to be a `RootRule`.

I'm going to experiment today with an automatic way to do that.

Reply:
IIRC, `ConsoleRuleTestBase` injects an `OptionsBootstrapper`, which allows for the creation of these subsystems.

I'll echo what I said on the black testing change PR: it's not clear to me that this level of boilerplate is worth it relative to a slightly higher level test.

Reply:
My opinion on this is that if we suffer the boilerplate to get to our goal of less heavy tests, it can also help us identify where boilerplate arises and motivate the cleanups that will make the boilerplate go away in the long turn.

Maybe as a review comment I would recommend a TODO linking to the appropriate github issue which tracks this particular boilerplate painpoint, and if there isn't an issue, this could be a good time to raise one.